src/charm.py

```diff
@@ -18,79 +17,54 @@
+    """ Raise this exception if one of the checks in main fails. """
+
+    def __init__(self, msg, status=None):
+        super().__init__()
+
+        self.msg = msg
+        self.status = status
 
+
+class Operator(CharmBase):
     def __init__(self, *args):
         super().__init__(*args)
-        if self.model.name != "kubeflow":
-            # Remove when this bug is resolved: https://github.com/kubeflow/kubeflow/issues/6136
-            self.model.unit.status = BlockedStatus(
-                "kubeflow-dashboard must be deployed to model named `kubeflow`:"
-                " https://git.io/J6d35"
-            )
-            return
 
-        if not self.unit.is_leader():
-            # We can't do anything useful when not the leader, so do nothing.
-            self.model.unit.status = WaitingStatus("Waiting for leadership")
-            return
         self.log = logging.getLogger(__name__)
         self.image = OCIImageResource(self, "oci-image")
 
-        try:
-            self.interfaces = get_interfaces(self)
-        except NoVersionsListed as err:
-            self.model.unit.status = WaitingStatus(str(err))
-            return
-        except NoCompatibleVersions as err:
-            self.model.unit.status = BlockedStatus(str(err))
-            return
-
         for event in [
             self.on.install,
             self.on.leader_elected,
             self.on.upgrade_charm,
             self.on.config_changed,
             self.on["kubeflow-profiles"].relation_changed,
+            self.on["ingress"].relation_changed,
         ]:
             self.framework.observe(event, self.main)
 
-        self.framework.observe(self.on["ingress"].relation_changed, self.configure_mesh)
-
-    def configure_mesh(self, event):
-        if self.interfaces["ingress"]:
-            self.interfaces["ingress"].send_data(
-                {
-                    "prefix": "/",
-                    "rewrite": "/",
-                    "service": self.model.app.name,
-                    "port": self.model.config["port"],
-                }
-            )
-
     def main(self, event):
         try:
-            image_details = self.image.fetch()
-        except OCIImageResourceError as e:
-            self.model.unit.status = e.status
-            self.log.info(e)
-            return
+            self._check_model_name()
 
-        model = self.model.name
+            self._check_leader()
 
-        if not (
-            (kf_profiles := self.interfaces["kubeflow-profiles"])
-            and kf_profiles.get_data()
-        ):
-            self.model.unit.status = WaitingStatus(
-                "Waiting for kubeflow-profiles relation data"
-            )
+            interfaces = self._get_interfaces()
+
+            image_details = self._check_image_details()
+
+            kf_profiles = self._check_kf_profiles(interfaces)
+        except CheckFailed as check_failed:
+            self.model.unit.status = check_failed.status(check_failed.msg)
```

Comment:
Related to above change to `CheckFailed`

```suggestion
        except CheckFailed as check_failed:
            self.model.unit.status = check_failed.status
```